### PR TITLE
vttablet: stream consolidation

### PR DESCRIPTION
**Pull request by vmg**
## Description

Happy Friday everyone! Here's the first iteration of a feature I believe is going to have a very significant impact in some of Vitess' largest deployments.

Here's the problem we're trying to solve: thundering herds. That's basically it. Sometimes databases suffer a usage pattern called a "thundering herd" where hundreds/thousands of queries happen at the same time, and all these queries are identical. Most of the time, this is caused either by a deployment, or by a cache that has been cleared or has expired.

Vitess already does a very good job at handling thundering herds in `vttablet` with our **query consolidation engine**, which prevents to identical point-queries from going to the upstream MySQL server: only one of the queries goes upstream, and all the others wait for the initial query to return. The issue is that our current consolidation engine **does not support streaming queries** -- understandable, because this is a hard problem to solve, but at the same time, this is particularly bad, because **streaming queries are by definition more expensive than point queries, so consolidating them is particularly beneficial to performance**. In practice, these queries often "block" one of our MySQL streaming connections for a long amount of time, and since we have a limited amount of connections from `vttablet` to MySQL, this results in a snowball effect where every single query in the herd takes longer than the previous one, because it needs to wait for all the previous queries to pick up a connection and release it back to the pool.

So, let's try to solve this problem: this PR implements a consolidation engine for streaming queries. I'm not sure whether this is a novel algorithm, but I haven't found any existing art on this so I had to cook it up myself. It was tricky to get right, and it's definitely complex, but the results seem worth it.

The consolidation engine for streaming queries is configurable with a given amount of memory, because it essentially acts as a cache. When a streaming query starts, the engine keeps track of it _and_ of all the intermediate rows it's returning from MySQL. If another identical stream shows up, it automatically catches up to the original stream with the data that has been cached in memory, and any new rows streamed from MySQL are fanned out to it. Once the original stream has grown large enough in memory, we clear its intermediate cache and new queries can no longer catch up to it, but all the existing followers keep receiving new rows as they come in.

In practice this works very well: with the default settings of 128MB total to be used for consolidation, and 2MB max for any consolidation stream, we can capture massive thundering herds even if they come with a latency differential of up to 250ms. The consolidation engine is smart, so when an identical query comes in that cannot catch up to the existing consolidation, it automatically promotes itself as a consolidation leader for any follow-up queries. A herd of 10.000 identical queries spread over 4s can be handled with only 4 MySQL streams, as opposed to, huh, 10.000.

The performance impact of this optimization is hard to graph, because the previous behavior is very pathological and the new behavior is very straightforward. As a sample:

consol

This is a thundering herd pattern that is performing an expensive query (1100 rows taking up to 2MB in total) from up to a thousand simultaneous connection. The latency pattern for the current version of `vttablet` (seen in orange) is all over the place: depending on the ordering that the scheduler gave to the incoming queries, the earliest queries take 30ms to finish, while the latest ones go all the way to 1.4 **seconds** -- this is because, again, every query needs to wait for all the previous ones to release the borrowed MySQL connections.

The consolidation engine (seen in blue), well, it consolidates the 1000 queries into a single stream, so they all finish simultaneously between 40 and 50ms.

This is it for now, I'm putting this up early for review/testing while I'm away next week for Easter. The code may have bugs, but the testing is comprehensive, with integration tests for all error behaviors (including clients that lag behind -- that was tricky to get right) and endtoend tests, with full `-race` coverage.

### Outstanding Issues

Nothing big. My main concern is that we're allocating more memory because the existing MySQL streaming client was reusing the rows of every Result set after yielding it (it's no longer safe to do this, because the result set is kept in memory temporarily to allow other queries to catch up). I've worked around the issue conservatively; an ideal approach would switch to MySQL connection to use a memory pool for rows reuse.

The other outstanding issue is that **the results of the stream callbacks are now immutable** -- understandably, as they're now shared between clients. I had to lift up a namespace-modification mutation so it didn't happen more than once between the consolidated queries, and the result was clean, but if somebody introduces another of these mutations in the future higher in the stack, it'll break things. The Go type system does not let us model this pointer as immutable because it uses type theory from 1973, so we'll have to keep an eye out to the race detector.

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by vmg** [comment]
Unfortunate hiccup: the CI server doesn't have enough concurrency to handle the 1000-client thundering herd in the stress tests, either with or without consolidation. I need to ponder this a bit; maybe the test needs to be simplified for CI altogether.

**Comment by dweitzman** [comment]
Seems like a generally good thing. The edge case I wonder about is what happens if two gates are caught up and streaming the same query, then one of them isn't able to receive results quickly (maybe the application is trashing and paging through results very slowly).

I haven't looked through the code, but I'm assuming the a single slow or stuck consolidated gate would fill up some buffer and then potentially block other gates from receiving results too?

If that's happening in production, seems like someone would be able to mitigate by setting ConsolidatorStreamTotalSize to 0 and restarting the tablets (to keep OLTP query consolidation and disable OLAP consolidation).

**Comment by vmg** [comment]
Back from my holiday!

> The edge case I wonder about is what happens if two gates are caught up and streaming the same query, then one of them isn't able to receive results quickly

This is explicitly handled and tested! Any clients that fall behind so much that they cannot catch up will eventually be disconnected with a specific error message. https://github.com/vitessio/vitess/pull/7752/files#diff-5fc9fba5ef5ce1a6a2744d9d42243b740ac8f76b29b06c16abcd805bb416673cR245-R282

**Comment by vmg** [comment]
The tests are now reliably green. It turned out to be quite hard to get these concurrent tests to always pass when run under the race detector, because it slows them down.

**Comment by vmg** [comment]
>  I've worked around the issue conservatively; an ideal approach would switch to MySQL connection to use a memory pool for rows reuse.

I've now implemented this feature in https://github.com/vitessio/vitess/pull/7752/commits/aa59b88e4497e53b1d0a66b60a29a40c2a8efb64, which should reduce memory allocations both when using and when not using the consolidator.

File: go/vt/vttablet/endtoend/stream_test.go [hunk 1]
```diff
@@ -41,6 +45,71 @@ func TestStreamUnion(t *testing.T) {
...
+			if err != nil {
+				return err
+			}
+		}
+		return client.Commit()
+	}
+
+	client := framework.NewClient()
+	err := populate(client)
+	if err != nil {
```
Comment (systay) [inline go/vt/vttablet/endtoend/stream_test.go:72]: nit: `require.NoError`

File: go/vt/vttablet/endtoend/stream_test.go [hunk 2]
```diff
@@ -41,6 +45,71 @@ func TestStreamUnion(t *testing.T) {
 	assert.Equal(t, 1, len(qr.Rows))
 }
 
+func TestStreamLongTail(t *testing.T) {
```
Comment (systay) [inline go/vt/vttablet/endtoend/stream_test.go:48]: Having a hard time following what you are testing here. Mind sprinkling comments to make it easier to follow?

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 3]
```diff
@@ -59,13 +60,39 @@ type QueryExecutor struct {
 	tabletType     topodatapb.TabletType
 }
 
+var streamResultPool = sync.Pool{New: func() interface{} {
+	return &sqltypes.Result{
+		Rows: make([][]sqltypes.Value, 0, 256),
```
Comment (systay) [inline go/vt/vttablet/tabletserver/query_executor.go:65]: nit: `256` could be extracted to a constant with a helpful name maybe?

File: go/vt/vttablet/tabletserver/stream_consolidator.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+	return &StreamConsolidator{
+		inflight:       make(map[string]*streamInFlight),
+		maxMemoryTotal: maxMemoryTotal,
+		maxMemoryQuery: maxMemoryQuery,
+		blocking:       false,
+		cleanup:        cleanup,
+	}
+}
+
+// StreamCallback is a function that yields every Result object from a streaming query
```
Comment (systay) [inline go/vt/vttablet/tabletserver/stream_consolidator.go:55]: I don't understand this comment. This function will be called with the result objects, it won't yield anything, right?

**Review by systay — APPROVED** [review]
Always a learning experience reading through your code :)

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 5]
```diff
@@ -245,6 +274,39 @@ func (qre *QueryExecutor) Stream(callback func(*sqltypes.Result) error) error {
...
+					defer dbConn.Recycle()
+					return qre.execStreamSQL(dbConn, sql, func(result *sqltypes.Result) error {
+						// this stream result is potentially used by more than one client, so
+						// the consolidator will return it to the pool once it knows it's no longer
+						// being shared
+
+						if replaceKeyspace != "" {
+							result.ReplaceKeyspace(replaceKeyspace)
+						}
+						return callback(result)
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go:304]: replace keyspace does not need to called always as not all queries will be for keyspace tables. It can be a information_schema query or _vt query.

There is similar check in `Execute` method where we check for if dbname is different from keyspace name and also if field.Database == dbname then only replace with the keyspace name.

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 6]
```diff
@@ -263,15 +325,20 @@ func (qre *QueryExecutor) Stream(callback func(*sqltypes.Result) error) error {
...
-	}
-	return qre.execStreamSQL(conn, sql, callback)
+	return qre.execStreamSQL(conn, sql, func(result *sqltypes.Result) error {
+		// this stream result is only used by the calling client, so it can be
+		// returned to the pool once the callback has fully returned
+		defer returnStreamResult(result)
+
+		if replaceKeyspace != "" {
+			result.ReplaceKeyspace(replaceKeyspace)
+		}
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go:335]: same as above.

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 5 above)
Comment (vmg) [reply go/vt/vttablet/tabletserver/query_executor.go:304]: @harshit-gangal this is the same behavior that we had before the PR; I've just refactored it by lifting it from the caller. I'm pretty sure we're only replacing the keyspace when it's required (see https://github.com/vitessio/vitess/pull/7752/files#diff-799d4762d9e64026cb623db2defeb53209df055a99c1106f535006a89e7b812aR282-R285, which I've lifted from https://github.com/vitessio/vitess/pull/7752/files#diff-28fc9df9d28d69ca625b85e18a6fa652d61bfaf6de03012ce2d8d0338363d70eL808-L817) Can you double-check?

Comment (vmg) [reply go/vt/vttablet/tabletserver/query_executor.go:304]: (unless the behavior before the PR was already broken, in that case I'll gladly fix it now :smile:)

Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/query_executor.go:304]: I realised that, it was broken before this PR
